utils/netutils/timeout_conn.go

```diff
@@ -51,3 +55,55 @@ func (tc *timeoutConn) SetWriteDeadline(t time.Time) error {
 	tc.writeTimeout = t.Sub(time.Now())
 	return tc.Conn.SetWriteDeadline(t)
 }
+
+// ReadFrom uses the underlying ReadFrom if available or does not use it if not available
+func (tc *timeoutConn) ReadFrom(r io.Reader) (n int64, err error) {
+	if wt, ok := r.(io.WriterTo); ok {
+		return wt.WriteTo(tc)
+	}
+	if rf, ok := tc.Conn.(io.ReaderFrom); ok {
+		var nn int64
+		for {
+			nn, err = rf.ReadFrom(io.LimitReader(r, maxSizeOfTransfer))
+			n += nn
+			if err != nil {
+				return
+			}
+			if nn != maxSizeOfTransfer {
+				return
+			}
+		}
+	}
+
+	// this is here because we need to write in smaller pieces in order to set the deadline
+	// directly using Copy will not use deadlines if used on the underlying net.Conn
+	// or will loop if used on timeoutConn directly
+	bufp := pool.Get().(*[]byte)
+	var readSize, writeSize int
+	var readErr error
+	for {
+		readSize, readErr = r.Read(*bufp)
+		n += int64(readSize)
+		writeSize, err = tc.Write((*bufp)[:readSize])
+		if err != nil {
+			return
+		}
+		if readSize != writeSize {
+			return n, io.ErrShortWrite
+		}
+		if readErr != nil {
+			pool.Put(bufp)
+			if err == io.EOF {
```

Comment:
I would guess you wanted to wrinte `if readErr == io.EOF`? When `err` is anything but `nil` this function is cut short in the 89th line.